pkg/ip/ip.go

```diff
@@ -220,3 +262,424 @@ func flipNthBit(ip *[]byte, bitNum uint) *[]byte {
+
+	var isIPv4 bool
+	var spanningMaskSize, bitLen, byteLen int
+	if lowest.To4() != nil {
+		isIPv4 = true
+		bitLen = ipv4BitLen
+		byteLen = net.IPv4len
+	} else {
+		bitLen = ipv6BitLen
+		byteLen = net.IPv6len
+	}
+
+	if isIPv4 {
+		spanningMaskSize = ipv4BitLen
+	} else {
+		spanningMaskSize = ipv6BitLen
+	}
+
+	// Convert to big Int so we can easily do bitshifting on the IP addresses,
+	// since golang only provides up to 64-bit unsigned integers.
+	lowestBig := big.NewInt(0).SetBytes(lowest)
+	highestBig := big.NewInt(0).SetBytes(highest)
+
+	// Starting from largest mask / smallest range possible, apply a mask one bit
+	// larger in each iteration to the upper bound in the range  until we have
+	// masked enough to pass the lower bound in the range. This
+	// gives us the size of the prefix for the spanning CIDR to return as
+	// well as the IP for the CIDR prefix of the spanning CIDR.
+	for spanningMaskSize > 0 && lowestBig.Cmp(highestBig) < 0 {
+		spanningMaskSize -= 1
+		mask := big.NewInt(1)
+		mask = mask.Lsh(mask, uint(bitLen-spanningMaskSize))
+		mask = mask.Mul(mask, big.NewInt(-1))
+		highestBig = highestBig.And(highestBig, mask)
+	}
+
+	// If ipv4, need to append 0s because math.Big gets rid of preceding zeroes.
+	if isIPv4 {
+		highest = append(ipv4LeadingZeroes, highestBig.Bytes()...)
+	} else {
+		highest = highestBig.Bytes()
+	}
+
+	// Int does not store leading zeroes.
+	if len(highest) == 0 {
+		highest = make([]byte, byteLen)
+	}
+
+	newNet := net.IPNet{IP: highest, Mask: net.CIDRMask(spanningMaskSize, bitLen)}
+	return newNet
+}
+
+type netWithRange struct {
+	First   *net.IP
+	Last    *net.IP
+	Network *net.IPNet
+}
+
+// CoalesceCIDRs transforms the provided list of CIDRs into the most-minimal equivalent set of CIDRs.
+// It removes CIDRs that are subnets of other CIDRs in the list, and groups together CIDRs that have the same mask size
+// into a CIDR of the same mask size provided that they share the same number of most significant mask-size bits.
+//
+// All IPs should be of the same type (IPv4, IPv6).
+func CoalesceCIDRs(cidrs []*net.IPNet) []*net.IPNet {
+
+	// TODO - check if all are IPv4 and IPv6
+	// Convert provided networks to netWithRange structures for manipulation
+	// during this function.
+	ranges := []*netWithRange{}
+
+	for _, network := range cidrs {
+		newNetToRange := ipNetToRange(*network)
+		ranges = append(ranges, &newNetToRange)
+	}
+
+PreLoop:
+	// Remove CIDRs which are contained within other CIDRs.
+	for j, cidr := range cidrs {
+		for i, cidr2 := range cidrs {
+			if i == j {
+				continue
+			}
+			if cidr.Contains(cidr2.IP) {
+				cidrs = append(cidrs[:i], cidrs[i+1:]...)
+				// Re-trigger loop since we have modified the slice we are iterating over.
+				goto PreLoop
+			}
+		}
+	}
+
+	// Sort the ranges. This sorts first by the last IP, then first IP, then by
+	// the IP network in the list itself
+	sort.Sort(NetsByRange(ranges))
+
+	// Merge adjacent CIDRs if possible.
```

Comment:
Would be nice to have it extracted as a function.

Reply:
Done.